### PR TITLE
fix watch=true doesn't respect cancellation token

### LINKED ISSUE #375: ListNamespacedServiceWithHttpMessagesAsync with watch=true doesn't respect cancellation token with empty list
When trying to watch for service based on fieldSelector and the service doesn't exist, cancellation token is not respected and the call hangs.

``` C#
var cts = new CancellationTokenSource();
cts.CancelAfter(TimeSpan.FromSeconds(30));

var service = await KubernetesClient.ListNamespacedServiceWithHttpMessagesAsync(
                        namespaceParameter: DefaultNamespace,
                        fieldSelector: "metadata.name=non-existing-service",
                        watch: true,
                        cancellationToken: cts.Token);
```

KuberneteClient version: `1.6.12`

_Edit: Added timeout for cancellation token cancellation._

### PR DESCRIPTION
**Pull request by tg123**
fix #375

### CODE REVIEW

File: src/KubernetesClient/WatcherDelegatingHandler.cs [hunk 1]
```diff
@@ -27,20 +27,76 @@ internal class WatcherDelegatingHandler : DelegatingHandler
...
+            private readonly Stream _innerStream;
+            private readonly CancellationToken _cancellationToken;
+
+            public CancelableStream(Stream innerStream, CancellationToken cancellationToken)
+            {
+                _innerStream = innerStream;
+                _cancellationToken = cancellationToken;
+            }
+
+            public override void Flush()
```
Comment (brendandburns) [inline src/KubernetesClient/WatcherDelegatingHandler.cs]: `public override void Flush() => _innertStream.Flush;` ?

File: src/KubernetesClient/WatcherDelegatingHandler.cs [hunk 2]
```diff
@@ -27,20 +27,76 @@ internal class WatcherDelegatingHandler : DelegatingHandler
...
+                _innerStream.Flush();
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                return _innerStream.ReadAsync(buffer, offset, count, _cancellationToken).ConfigureAwait(false)
+                    .GetAwaiter().GetResult();
+            }
+
+            public override long Seek(long offset, SeekOrigin origin)
```
Comment (brendandburns) [inline src/KubernetesClient/WatcherDelegatingHandler.cs]: Can you use delegate syntax here?

File: src/KubernetesClient/WatcherDelegatingHandler.cs [hunk 3]
```diff
@@ -49,13 +105,13 @@ protected override async Task SerializeToStreamAsync(Stream stream, TransportCon
...
 
-                StreamReader = new PeekableStreamReader(_originStream);
+                StreamReader = new PeekableStreamReader(new CancelableStream(_originStream, _cancellationToken));
 
                 var firstLine = await StreamReader.PeekLineAsync();
 
                 var writer = new StreamWriter(stream);
 
-//                using (writer) // leave open
+                //                using (writer) // leave open
```
Comment (brendandburns) [inline src/KubernetesClient/WatcherDelegatingHandler.cs]: let's just delete this line.

**Comment by brendandburns** [comment]
Couple of small style comments.

File: src/KubernetesClient/WatcherDelegatingHandler.cs [hunk 4]
```diff
@@ -27,20 +27,63 @@ internal class WatcherDelegatingHandler : DelegatingHandler
...
+            public CancelableStream(Stream innerStream, CancellationToken cancellationToken)
+            {
+                _innerStream = innerStream;
+                _cancellationToken = cancellationToken;
+            }
+
+            public override void Flush() => _innerStream.Flush();
+
+            public override int Read(byte[] buffer, int offset, int count) =>
+                _innerStream.ReadAsync(buffer, offset, count, _cancellationToken).ConfigureAwait(false)
```
Comment (andykernahan) [inline src/KubernetesClient/WatcherDelegatingHandler.cs]: Consider omitting `ConfigureAwait(false)` as it does not affect the behaviour of the awaiter's `GetResult`. It is also potentially misleading to the reader as they may falsely believe these methods (`CancelableStream.{Read, Write}`) to be insusceptible to the classic async/await deadlock.

Comment (andykernahan) [inline src/KubernetesClient/WatcherDelegatingHandler.cs]: Consider omitting `ConfigureAwait(false)` as it does not affect the behaviour of the awaiter's `GetResult`; therefore, it is potentially misleading to the reader as they may falsely believe these methods (`CancelableStream.{Read, Write}`) to be insusceptible to the classic async/await deadlock.

Comment (tg123) [reply src/KubernetesClient/WatcherDelegatingHandler.cs]: fixed thanks

**Comment by brendandburns** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**